app/presenters/paginator.presenter.js

```diff
@@ -0,0 +1,242 @@
+ *
+ *  `[1] 2 3 4 5 6 7 Next -->`
+ *
+ * ### Complex
+ *
+ * This applies where we have 8 or more pages of results. It means at least one page item we will need to be an
+ * ellipsis. The next problem is determining if both should be displayed, or only one and if only one where?
+ *
+ * To do this we break the complex component down into a further 3 types; start, middle and end. They are determined
+ * based on which is the current page.
+ *
+ * #### Start
+ *
+ * If the current page is one of the first 4 pages we define the pagination type as 'complex start'. If, for example,
+ * the current page is `[2]` and we took the guidance at face value we could have generated the paginator as
+ *
+ * `<- Previous 1 [2] 3 ... 42 Next ->`
+ *
+ * This control only displays 5 page items. But once the current page is `[5]` or more we are required to show ellipsis
+ * at both ends.
+ *
+ * `<- Previous 1 ... 4 [5] 6 ... 42 Next ->`
+ *
+ * Now we're displaying 7 page items. Our approach removes the inconsistency and always shows 7 page items. This means
+ * when `[2]` is the current page the paginator will display.
+ *
+ * `<- Previous 1 [2] 3 4 5 ... 42 Next ->`
+ *
+ * #### Middle
+ *
+ * This applies where the current page is greater than 4 and less than the number of pages minus 4 (for example, if the
+ * number of pages is 42 this means the current page is greater than 4 and less than 39). If it is we define the
+ * pagination type as 'complex middle'.
+ *
+ * When this is the case the selected page is in the 'middle' and an ellipsis needs to be shown at both ends.
+ *
+ * `<- Previous 1 ... 4 [5] 6 ... 42 Next ->`
+ *
+ * ##### End
+ *
+ * When the current page is one of the last 4 pages we define the pagination type as 'complex end'. If, for example,
+ * the number of pages is 42 and the current page is 39 we'll generate the following paginator.
+ *
+ * `<- Previous 1 ... 38 [39] 50 41 42 Next ->`
+ *
+ * @param {number} numberOfRecords - the total number of records or results of the thing being paginated
+ * @param {number} selectedPageNumber - the page of results selected for viewing
+ *
+ * @returns {Object} if no pagination is needed just the `numberOfPages: 1` is returned else a `component:` property is
+ * also included that can be directly passed to the `govukPagination()` in the view.
+ */
+function go (numberOfRecords, selectedPageNumber) {
+  const numberOfPages = Math.ceil(numberOfRecords / DatabaseConfig.defaultPageSize)
+
+  if (numberOfPages === 1) {
+    return { numberOfPages }
+  }
+
+  const component = _component(selectedPageNumber, numberOfPages)
+
+  return {
+    component,
+    numberOfPages
+  }
+}
+
+function _component (selectedPageNumber, numberOfPages) {
+  const items = _items(selectedPageNumber, numberOfPages)
+
+  const component = { items }
+
+  if (selectedPageNumber !== 1) {
+    component.previous = { href: `/system/bill-runs?page=${selectedPageNumber - 1}` }
```

Comment:
More of a question really. If the intention is to reuse this presenter on other pages that require this component, should we be hardcoding the URL `/system/bill-runs`? I guess currently it's fine as it is the only page that uses this presenter.  But is it worth setting up the URL as an additional parameter now, or is that something to do if/when we need to?

Reply:
🎯 Brilliant spot. Yes, I should be making this a param.